src/protocol_select/mod.rs

```diff
@@ -0,0 +1,333 @@
+};
+
+use bytes::Bytes;
+use flatbuffers::{get_root, FlatBufferBuilder};
+use futures::{future, prelude::*};
+use log::debug;
+use std::{collections::HashMap, io};
+use tokio::codec::{length_delimited::LengthDelimitedCodec, Framed};
+use tokio::prelude::{AsyncRead, AsyncWrite};
+
+#[rustfmt::skip]
+#[allow(clippy::all)]
+mod protocol_select_generated;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub(crate) struct ProtocolMessage {
+    pub name: String,
+    pub support_versions: Vec<String>,
+}
+
+impl ProtocolMessage {
+    pub fn new(name: &str, support_versions: Vec<String>) -> Self {
+        ProtocolMessage {
+            name: name.to_owned(),
+            support_versions,
+        }
+    }
+
+    pub fn encode(&self) -> Vec<u8> {
+        let mut fbb = FlatBufferBuilder::new();
+        let name = fbb.create_string(&self.name);
+        let versions = &self
+            .support_versions
+            .iter()
+            .map(|version| fbb.create_string(version))
+            .collect::<Vec<_>>();
+        let versions = fbb.create_vector(versions);
+
+        let mut builder = ProtocolMessageBuilder::new(&mut fbb);
+        builder.add_name(name);
+        builder.add_support_versions(versions);
+        let data = builder.finish();
+
+        fbb.finish(data, None);
+        fbb.finished_data().to_vec()
+    }
+
+    pub fn decode(data: &[u8]) -> Result<Self, ()> {
+        let fbs_protocol_message = get_root::<FBSProtocolMessage>(data);
+        if fbs_protocol_message.name().is_none()
+            || fbs_protocol_message.support_versions().is_none()
+        {
+            Err(())
+        } else {
+            let mut versions: Vec<String> = Vec::new();
+            let fbs_versions = fbs_protocol_message.support_versions().unwrap();
+            for i in 0..fbs_versions.len() {
+                versions.push(fbs_versions.get(i).to_owned());
+            }
+            Ok(ProtocolMessage {
+                name: fbs_protocol_message.name().unwrap().to_owned(),
+                support_versions: versions,
+            })
+        }
+    }
+}
+
+/// Performs a handshake on the given socket.
+///
+/// Select the protocol version, return a handle that implements the `AsyncWrite` and `AsyncRead` trait,
+/// plus the protocol name, plus the version option.
+pub(crate) fn client_select<T: AsyncWrite + AsyncRead + Send>(
+    handle: T,
+    message: ProtocolMessage,
+) -> impl Future<Item = (T, String, Option<String>), Error = io::Error> {
+    let socket = Framed::new(handle, LengthDelimitedCodec::new());
+    future::ok::<_, io::Error>(message)
+        .and_then(|message| {
+            socket
+                .send(Bytes::from(message.encode()))
+                .from_err()
+                .map(|socket| socket)
+        })
+        .and_then(|socket| {
+            socket
+                .into_future()
+                .map_err(|(e, socket)| {
+                    let _ = socket.into_inner().shutdown();
+                    e
```

Comment:
Why `shutdown()` here?

Reply:
Any error occurs during the handshake, must close the sub stream